Tilt setup for kpng

ksankeerth opened the pull request:
This PR add tilt setup for kpng. 

**How it works?**
eg:
```console
root@realistic-planarian:/home/ubuntu/kpng# make tilt-setup i=ipv4 b=nft

root@realistic-planarian:/home/ubuntu/kpng# make tilt-up

````
`make tilt-setup` will create a cluster and install tools such as kubectl, bpf2go, and kind. It'll also create a file called `tilt.env` with the environment variables (including ip family, backend, and etc).

`make tilt-up` will run `tilt up` command and the environment variables set previously will be loaded into tilt session.  We will generate kpng Daemonset yaml based on the environment variables.

To change the backend, We can modify  the file `tilt.env`. It'll trigger an update to tilt. We can also trigger an update from the Web UI as well. 

Currently, Tilt is not configured build and re-deploy on change of source code but it can be done. I avoided that it may not be effective to run tilt update every code changes.

Appreciate your feedback.

Fixes #383

knabben commented:
@ksankeerth I was thinking about tilt last week and the adoption, the hack folder has a few scripts to help in the development workflow (and everyone already uses). What do you think about doing a demo on Friday so ppl are on the same page in the devxp

knabben commented on Tiltfile:
```diff
@@ -0,0 +1,21 @@
+version_settings(constraint='>=0.22.2')
+
+load('ext://dotenv', 'dotenv')
+dotenv('tilt.env')
```
do we have a default tilt.env for development?

ksankeerth replied:
`tilt.env` is not required for `tilt`.  

When we use `kind` with `tilt`, the tilt expects the cluster to be ready. So  I added `make tilt-setup i=ipv4 b=nft` for creating a cluster. When we create a kind cluster, We need to know  the `ip family` to decide `podSubnet` and  `serviceSubnet`

When we run `tilt up` or `make tilt-up`, We need to generate kpng DaemonSet `yaml` and build image. then deploy into the kind cluster.  To generate daemonset yaml, We need to know what the ip family and backend we provided when setup the cluster. So I use `tilt.env` file to store the environment variables including backend, and ip-family(`tilt.env` will be created when we run `make tilt-setup`).  Then we can read it inside tilt session. 

If we need to change the backend, We can change it in `tilt.env` file. It'll trigger an update to `tilt`.

There may be a better approach than this 🙂

knabben replied:
what about a tilt.dev.example and say in the readme you can copy it over. so the user has a template (unless make tilt-up) is generating this automatically

dougsland replied:
+1

dougsland replied:
@knabben @ksankeerth just wondering if the Tiltfile should be inside the tilt dir?

ksankeerth replied:
@dougsland I think it's a good suggestion. Mostly, the users don't need to change Tiltfile. I'll move to tilt dir.

knabben commented on hack/tilt/deploy.sh:
```diff
@@ -0,0 +1,97 @@
+BIN_DIR=$(dirname $SCRIPT_DIR)/temp/tilt/bin 
+
+# system data
+NAMESPACE="kube-system"
```
should these be in the tilt.dev instead

ksankeerth replied:
We'll use this namespace to deploy `kpng` DaemonSet. We may change it to `tilt.dev`

dougsland replied:
+1

knabben commented on hack/tilt/setup.sh:
```diff
@@ -0,0 +1,508 @@
+SERVICE_CLUSTER_IP_RANGE_V6="fd6d:706e:6702::/112"
+KPNG_SERVER_ADDRESS="unix:///k8s/proxy.sock"
+KPNG_DEBUG_LEVEL=4
+KIND="kindest/node:v1.22.13@sha256:4904eda4d6e64b402169797805b8ec01f50133960ad6c19af45173a27eadf959"
```
same for these vars, should be on tilt.env

dougsland replied:
+1

ksankeerth replied:
Currently, only the below variable is in `setup.sh`. others were moved to `common.sh`.
```console
KIND="kindest/node:v1.22.13@sha256:4904eda4d6e64b402169797805b8ec01f50133960ad6c19af45173a27eadf959"

```

knabben commented on hack/tilt/setup.sh:
```diff
@@ -0,0 +1,508 @@
+    esac
+}
+
+function install_binaries {
```
it seems a lot of copy and paste from test_e2e.sh. I rather avoid this and reuse the functions from there, maybe split that file on a common.sh

ksankeerth replied:
totally agree with you. I'll separate common functions into a file.

knabben commented on hack/tilt/deploy.sh:
```diff
@@ -0,0 +1,97 @@
+
+export_vars 
+
+go run ${SCRIPT_DIR}/kpng-ds-yaml-gen.go ${SCRIPT_DIR}/kpng-deployment-ds-template.txt ${SCRIPT_DIR}/kpng-deployment-ds.yaml
```
at this point I'm afraid of scripts doing the same thing without reuse :(

ksankeerth replied:
yes, I'll try to move common codes to a separate file

ksankeerth commented:
> @ksankeerth I was thinking about tilt last week and the adoption, the hack folder has a few scripts to help in the development workflow (and everyone already uses). What do you think about doing a demo on Friday so ppl are on the same page in the devxp

Sure, I'll join next meeting

ksankeerth commented:
@knabben Thanks for the review and suggestions. Not sure you prefer having tilt.env file #395 (comment). I couldn't find a better approach than this

knabben commented:
> @knabben Thanks for the review and suggestions. Not sure you prefer having tilt.env file #395 (comment). I couldn't find a better approach than this

maybe bring tilt.env.example with the defaults, and write some notes in the README do copy it.

ksankeerth commented:
> > @knabben Thanks for the review and suggestions. Not sure you prefer having tilt.env file #395 (comment). I couldn't find a better approach than this
> 
> maybe bring tilt.env.example with the defaults, and write some notes in the README do copy it.

@knabben  didn't have chance to do demo on last Friday.  I have update the branch with the below changes.

1. moved common functions to common.sh 
2. added instructions to use `Tilt` in README.md 
3. added comments for make commands to show them in `make help`

Please review the changes.

I  recorded how this Tilt setup works for KPNG. You may look at it https://drive.google.com/file/d/1x6b00nef2KTEKEOj4Q8HEuerxompYWmn/view?usp=share_link

Thanks

knabben commented on Tiltfile:
```diff
@@ -0,0 +1,21 @@
+    delete_cmd='echo "[tilt] deployment completed"',
+    deps=['tilt.env']
+)
+
```
i think you dont need these multiple lines

knabben commented on hack/tilt/setup.sh:
```diff
@@ -0,0 +1,239 @@
+    printf "Both of 'i' and 'b' must be specified.\n"
+    help
+fi
+
```
same here with the newlines

knabben commented:
/lgtm

thanks for recording the demo, I think tilt is a good improvement.

dougsland commented on hack/common.sh:
```diff
@@ -0,0 +1,386 @@
+}
+
+
+
```
no need extra lines here

ksankeerth replied:
> Another question:
> 
> `$ make tilt-setup i=ipv4 b=iptables`
> 
> If it's used iptables as backend, is it worth installing bpg2to ?
> 
> ```
> [ INFO ] 'bpf2go' not found, installing with 'go'
> go: downloading github.com/cilium/ebpf v0.9.2
> go: downloading golang.org/x/sys v0.0.0-20220829200755-d48e67d00261
> # internal/race
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # internal/unsafeheader
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # runtime/internal/sys
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # internal/goexperiment
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # unicode/utf8
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # internal/itoa
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # math/bits
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # unicode
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # encoding
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # golang.org/x/sys/internal/unsafeheader
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # internal/abi
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # sync/atomic
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # internal/cpu
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # runtime/internal/atomic
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> [ FAILED ] cannot install bpf2go
> ```

@dougsland Currently all the setup are done when running `make tilt-setup`.  We only build images and redeploy once the `Tilt` server is started.   If we install `bpf2go` only when users try to use `ebpf` backend , We may have to increase `Tilt` timeout as well. If the `Tilt` was not able to trigger an re-deployment within the `timeout`, it'll fail.

I think the expectation (as per https://github.com/kubernetes-sigs/kpng/issues/383#issue-1410480330)  is that user will start with one backend and later they'll change.

dougsland commented on hack/tilt/setup.sh:
```diff
@@ -0,0 +1,239 @@
+- role: control-plane
+- role: worker
+- role: worker
+EOF
```
I would recommend for a future patch to move this kind config to a external file and inside the this script execute a sed. Just trying to make it easier and not huge file.

ksankeerth replied:
I'll submit a follow-up PR after this.

dougsland commented:
OKay, looks nice one here, specially as moving stuff to common.sh instead of re-writing the wheel, well done. I just made a few comments.

Another nit comment: I would add the demo as embedded in the tilt/README.md

dougsland commented:
Another question:

```$ make tilt-setup  i=ipv4 b=iptables```

If it's used iptables as backend, is it worth installing bpg2to ?

```
[ INFO ] 'bpf2go' not found, installing with 'go'
go: downloading github.com/cilium/ebpf v0.9.2
go: downloading golang.org/x/sys v0.0.0-20220829200755-d48e67d00261
# internal/race
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# internal/unsafeheader
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# runtime/internal/sys
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# internal/goexperiment
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# unicode/utf8
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# internal/itoa
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# math/bits
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# unicode
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# encoding
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# golang.org/x/sys/internal/unsafeheader
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# internal/abi
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# sync/atomic
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# internal/cpu
compile: version "go1.17.6" does not match go tool version "go1.18.8"
# runtime/internal/atomic
compile: version "go1.17.6" does not match go tool version "go1.18.8"
[ FAILED ] cannot install bpf2go
```

ksankeerth commented:
Thanks @dougsland and @knabben for the reviews. I'll address your comments and questions.

jayunit100 commented:
this is awesome thanks lets merge and iterate once non-blocking comments are done !!!!!!!!!!!!

ksankeerth commented:
> Another question:
> 
> `$ make tilt-setup i=ipv4 b=iptables`
> 
> If it's used iptables as backend, is it worth installing bpg2to ?
> 
> ```
> [ INFO ] 'bpf2go' not found, installing with 'go'
> go: downloading github.com/cilium/ebpf v0.9.2
> go: downloading golang.org/x/sys v0.0.0-20220829200755-d48e67d00261
> # internal/race
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # internal/unsafeheader
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # runtime/internal/sys
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # internal/goexperiment
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # unicode/utf8
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # internal/itoa
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # math/bits
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # unicode
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # encoding
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # golang.org/x/sys/internal/unsafeheader
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # internal/abi
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # sync/atomic
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # internal/cpu
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> # runtime/internal/atomic
> compile: version "go1.17.6" does not match go tool version "go1.18.8"
> [ FAILED ] cannot install bpf2go
> ```

@dougsland  Currently all the setup are done when running `make tilt-setup`.  We only build images and redeploy once the `Tilt` server is started.   If we install `bpf2go` only when users try to use `ebpf` backend , We may have to increase `Tilt` timeout as well. If the `Tilt` was not able to trigger an re-deployment within the `timeout`, it'll fail.

I think the expectation (as per https://github.com/kubernetes-sigs/kpng/issues/383#issue-1410480330)  is that user will start with one backend and later they'll change.

ksankeerth commented:
@dougsland @knabben I have addressed most of your comments and updated the branch. please take a look at it.

knabben commented:
this is very cool thanks @ksankeerth 

/lgtm
/approve

jayunit100 commented:
/lgtm
/approve
